// @aws-cdk/aws-ecs-patterns/test/ec2/integ.scheduled-ecs-task.lit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ec2 = require("@aws-cdk/aws-ec2");
const ecs = require("@aws-cdk/aws-ecs");
const events = require("@aws-cdk/aws-events");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../../lib");
const app = new cdk.App();
class EventStack extends cdk.Stack {
    constructor(scope, id) {
        super(scope, id);
        const vpc = new ec2.Vpc(this, 'Vpc', { maxAzs: 1 });
        const cluster = new ecs.Cluster(this, 'EcsCluster', { vpc });
        cluster.addCapacity('DefaultAutoScalingGroup', {
            instanceType: new ec2.InstanceType('t2.micro')
        });
        /// !show
        // Create the scheduled task
        new lib_1.ScheduledEc2Task(this, 'ScheduledEc2Task', {
            cluster,
            image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
            desiredTaskCount: 2,
            memoryLimitMiB: 512,
            cpu: 1,
            environment: { TRIGGER: 'CloudWatch Events' },
            schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
        });
        /// !hide
    }
}
new EventStack(app, 'aws-ecs-integ-ecs');
app.synth();
