// @aws-cdk/aws-events-targets/test/sns/sns.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
const events = require("@aws-cdk/aws-events");
const sns = require("@aws-cdk/aws-sns");
const core_1 = require("@aws-cdk/core");
const targets = require("../../lib");
test('sns topic as an event rule target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const topic = new sns.Topic(stack, 'MyTopic');
    const rule = new events.Rule(stack, 'MyRule', {
        schedule: events.Schedule.rate(core_1.Duration.hours(1)),
    });
    // WHEN
    rule.addTarget(new targets.SnsTopic(topic));
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::SNS::TopicPolicy', {
        PolicyDocument: {
            Statement: [
                {
                    Sid: "0",
                    Action: "sns:Publish",
                    Effect: "Allow",
                    Principal: { Service: { "Fn::Join": ["", ["events.", { Ref: "AWS::URLSuffix" }]] } },
                    Resource: { Ref: "MyTopic86869434" }
                }
            ],
            Version: "2012-10-17"
        },
        Topics: [{ Ref: "MyTopic86869434" }]
    }));
    assert_1.expect(stack).to(assert_1.haveResource('AWS::Events::Rule', {
        ScheduleExpression: "rate(1 hour)",
        State: "ENABLED",
        Targets: [
            {
                Arn: { Ref: "MyTopic86869434" },
                Id: "Target0"
            }
        ]
    }));
});
test('multiple uses of a topic as a target results in a single policy statement', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const topic = new sns.Topic(stack, 'MyTopic');
    // WHEN
    for (let i = 0; i < 5; ++i) {
        const rule = new events.Rule(stack, `Rule${i}`, {
            schedule: events.Schedule.rate(core_1.Duration.hours(1)),
        });
        rule.addTarget(new targets.SnsTopic(topic));
    }
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::SNS::TopicPolicy', {
        PolicyDocument: {
            Statement: [
                {
                    Action: "sns:Publish",
                    Effect: "Allow",
                    Principal: { Service: { "Fn::Join": ["", ["events.", { Ref: "AWS::URLSuffix" }]] } },
                    Resource: { Ref: "MyTopic86869434" },
                    Sid: "0"
                }
            ],
            Version: "2012-10-17"
        },
        Topics: [{ Ref: "MyTopic86869434" }]
    }));
});
