// @aws-cdk/aws-ecs-patterns/test/fargate/integ.scheduled-fargate-task.lit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ec2 = require("@aws-cdk/aws-ec2");
const ecs = require("@aws-cdk/aws-ecs");
const events = require("@aws-cdk/aws-events");
const cdk = require("@aws-cdk/core");
const path = require("path");
const lib_1 = require("../../lib");
const app = new cdk.App();
class EventStack extends cdk.Stack {
    constructor(scope, id) {
        super(scope, id);
        const vpc = new ec2.Vpc(this, 'Vpc', { maxAzs: 1 });
        const cluster = new ecs.Cluster(this, 'FargateCluster', { vpc });
        // Create the scheduled task
        new lib_1.ScheduledFargateTask(this, 'ScheduledFargateTask', {
            cluster,
            image: new ecs.AssetImage(path.join(__dirname, '..', 'demo-image')),
            desiredTaskCount: 2,
            memoryLimitMiB: 512,
            cpu: 256,
            environment: { TRIGGER: 'CloudWatch Events' },
            schedule: events.Schedule.rate(cdk.Duration.minutes(2)),
        });
    }
}
new EventStack(app, 'aws-fargate-integ');
app.synth();
