// @aws-cdk/aws-ecs-patterns/lib/ecs/load-balanced-ecs-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const aws_ecs_1 = require("@aws-cdk/aws-ecs");
const load_balanced_service_base_1 = require("../base/load-balanced-service-base");
/**
 * A single task running on an ECS cluster fronted by a load balancer
 */
class LoadBalancedEc2Service extends load_balanced_service_base_1.LoadBalancedServiceBase {
    /**
     * Constructs a new instance of the LoadBalancedEc2Service class.
     */
    constructor(scope, id, props) {
        super(scope, id, props);
        const taskDefinition = new aws_ecs_1.Ec2TaskDefinition(this, 'TaskDef', {
            executionRole: props.executionRole,
            taskRole: props.taskRole
        });
        const containerName = props.containerName !== undefined ? props.containerName : 'web';
        const container = taskDefinition.addContainer(containerName, {
            image: props.image,
            cpu: props.cpu,
            memoryLimitMiB: props.memoryLimitMiB,
            memoryReservationMiB: props.memoryReservationMiB,
            environment: props.environment,
            secrets: props.secrets,
            logging: this.logDriver,
        });
        container.addPortMappings({
            containerPort: props.containerPort || 80
        });
        this.service = new aws_ecs_1.Ec2Service(this, "Service", {
            cluster: this.cluster,
            desiredCount: this.desiredCount,
            taskDefinition,
            assignPublicIp: this.assignPublicIp,
            serviceName: props.serviceName,
            healthCheckGracePeriod: props.healthCheckGracePeriod,
        });
        this.addServiceAsTarget(this.service);
    }
}
exports.LoadBalancedEc2Service = LoadBalancedEc2Service;
