// @aws-cdk/aws-events-targets/lib/lambda.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const iam = require("@aws-cdk/aws-iam");
/**
 * Use an AWS Lambda function as an event rule target.
 */
class LambdaFunction {
    constructor(handler, props = {}) {
        this.handler = handler;
        this.props = props;
    }
    /**
     * Returns a RuleTarget that can be used to trigger this Lambda as a
     * result from a CloudWatch event.
     */
    bind(rule, _id) {
        const permissionId = `AllowEventRule${rule.node.uniqueId}`;
        if (!this.handler.permissionsNode.tryFindChild(permissionId)) {
            this.handler.addPermission(permissionId, {
                action: 'lambda:InvokeFunction',
                principal: new iam.ServicePrincipal('events.amazonaws.com'),
                sourceArn: rule.ruleArn
            });
        }
        return {
            id: '',
            arn: this.handler.functionArn,
            input: this.props.event,
            targetResource: this.handler,
        };
    }
}
exports.LambdaFunction = LambdaFunction;
